sql/015-operators-eq.sql

```diff
@@ -0,0 +1,525 @@
+-- Operators for unique comparisons of cs_encrypted_v1 types
+--
+-- Support for the following comparisons:
+--
+--      cs_encrypted_v1 = cs_encrypted_v1
+--      cs_encrypted_v1 <> cs_encrypted_v1
+--      cs_encrypted_v1 = jsonb
+--      cs_encrypted_v1 <> jsonb
+--      cs_encrypted_v1 = text
+--      cs_encrypted_v1 <> text
+--
+
+DROP OPERATOR IF EXISTS = (cs_encrypted_v1, cs_encrypted_v1);
+DROP FUNCTION IF EXISTS cs_encrypted_eq_v1(a cs_encrypted_v1, b cs_encrypted_v1);
+
+CREATE FUNCTION cs_encrypted_eq_v1(a cs_encrypted_v1, b cs_encrypted_v1)
+  RETURNS boolean
+  IMMUTABLE STRICT PARALLEL SAFE
+AS $$
+  DECLARE
+    u boolean;
+    o boolean;
+  BEGIN
+    BEGIN
+      u := (SELECT cs_unique_v1(a) = cs_unique_v1(b));
```

Comment:
I'm not 100% what the threat model might be but its worth noting that this comparison won't run in constant time. It _might_ leave the possibility for an attacker to learn the value of the plaintext in certain circumstances.

There's a constant time comparison function in the EQL repo somewhere I think.

Reply:
Something like this: https://github.com/cipherstash/encrypt-query-language/blob/2ee50af4abcc68e5346baa89687cf893e67daf89/sql/001-ore-cllw.sql#L21

Reply:
`cs_unique_v1` is the existing function, and simply extracts the index value from the `eql::ciphertext`

Reply:
Unique indexes are cast as `text`, so this ends up using the default text eq operators.

Reply:
Yeah that's what I figured. The standard equality operator will bail as soon as a character differs which can lead to timing side-channel attacks. A constant time comparison should always take the same amount of time regardless of which character index differs first.

Reply:
Oh right. We should fix that for unique in general then